### PR TITLE
Generate new certs to be used for joining nodes

### PR DESCRIPTION
**Pull request by divolgin**

### CODE REVIEW

File: kotsadm/pkg/k8s/cluster.go [hunk 1]
```diff
@@ -42,22 +44,26 @@ func GenerateBootstrapToken(client kubernetes.Interface, ttl time.Duration) (str
...
+
+	config, err := kubeadmconfig.DefaultedInitConfiguration(&kubeadmapiv1beta2.InitConfiguration{}, &kubeadmapiv1beta2.ClusterConfiguration{})
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get kotsadm config")
 	}
 
 	key, err := copycerts.CreateCertificateKey()
 	if err != nil {
-		return "", err
+		return "", errors.Wrap(err, "failed to genertae key")
```
Comment (marccampbell) [inline kotsadm/pkg/k8s/cluster.go:60]: spelling

File: kotsadm/pkg/kurl/join_cert.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+		status, err := client.CoreV1().Pods(pod.Namespace).Get(ctx, pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return "", errors.Wrap(err, "failed to get pod")
+		}
+		if status.Status.Phase == corev1.PodRunning ||
+			status.Status.Phase == corev1.PodFailed ||
+			status.Status.Phase == corev1.PodSucceeded {
+			break
+		}
+		time.Sleep(time.Second * 1)
```
Comment (marccampbell) [inline kotsadm/pkg/kurl/join_cert.go:49]: this could hang forever, wdyt about adding a timeout in here?

File: kotsadm/pkg/kurl/join_cert.go [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+	return key, nil
+}
+
+func getCertGenLogs(ctx context.Context, client kubernetes.Interface, pod *corev1.Pod) ([]byte, error) {
+	// maxLines := int64(10000)
+	podLogOpts := corev1.PodLogOptions{
+		Follow:    false,
+		Container: pod.Spec.Containers[0].Name,
+		// TailLines:
```
Comment (marccampbell) [inline kotsadm/pkg/kurl/join_cert.go:70]: want to delete?

File: kotsadm/pkg/kurl/join_cert.go [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+	_, err = io.Copy(buf, podLogs)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to copy log")
+	}
+
+	return buf.Bytes(), nil
+}
+
+func parseCertGenOutput(logs []byte) (string, error) {
+	// Output looks like this:
```
Comment (marccampbell) [inline kotsadm/pkg/kurl/join_cert.go:91]: yikes, there's no JSON output or less brittle?

File: kotsadm/pkg/kurl/join_cert.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+						HostPath: &corev1.HostPathVolumeSource{
+							Path: "/usr/bin/kubeadm",
+							Type: &binVolumeType,
+						},
+					},
+				},
+			},
+			Containers: []corev1.Container{
+				{
+					Image:           existingDeployment.Spec.Template.Spec.Containers[apiContainerIndex].Image,
```
Comment (marccampbell) [inline kotsadm/pkg/kurl/join_cert.go:178]: :thinking: the effort to get the container image is appreciated but also seems hard and potentially fragile. 

I am assuming this is needed for airgap installs? Can't we just look for the kurl config map / image pull secret with a label or some other methods to look it up and construct it? Maybe not, this isn't available in the downward api either.

No objections here, but it feels tricky and brittle.

File: kotsadm/pkg/kurl/configmap.go [hunk 6]
```diff
@@ -69,7 +69,7 @@ func UpdateConfigMap(client kubernetes.Interface, generateBootstrapToken, upload
 	if uploadCerts {
 		certsDuration := time.Hour * 2
 		certsExpiration := time.Now().Add(certsDuration)
-		key, err := k8s.UploadCertsWithNewKey(client)
+		key, err := createCertAndKey(context.TODO(), client, "default")
```
Comment (emosbaugh) [inline kotsadm/pkg/kurl/configmap.go:72]: should you use `os.Getenv("POD_NAMESPACE")` here?

File: kotsadm/pkg/kurl/join_cert.go (hunk 2 above)
Comment (emosbaugh) [reply kotsadm/pkg/kurl/join_cert.go:49]: better yet use the context with timeout

```go
if err := ctx.Err(); err != nil {
  return err
}
```

File: kotsadm/pkg/kurl/join_cert.go (hunk 4 above)
Comment (emosbaugh) [reply kotsadm/pkg/kurl/join_cert.go:91]: it seems like the other option would be to write this in go and import kotsadm lib?

File: kotsadm/pkg/kurl/join_cert.go [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+			Kind:       "Pod",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: existingDeployment.Namespace,
+			Labels:    existingDeployment.Labels,
+		},
+		Spec: corev1.PodSpec{
+			SecurityContext: &securityContext,
+			NodeSelector: map[string]string{
```
Comment (emosbaugh) [inline kotsadm/pkg/kurl/join_cert.go:151]: i think you should add toleration here as well for the future

```
      tolerations:
      - key: node-role.kubernetes.io/master
        effect: NoSchedule
```

File: kotsadm/pkg/kurl/join_cert.go [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+					},
+				},
+			},
+			Containers: []corev1.Container{
+				{
+					Image:           existingDeployment.Spec.Template.Spec.Containers[apiContainerIndex].Image,
+					ImagePullPolicy: corev1.PullNever,
+					Name:            "join-cert-gen",
+					Command:         []string{"/usr/bin/kubeadm"},
+					Args:            []string{"init", "phase", "upload-certs", "--upload-certs"},
```
Comment (emosbaugh) [inline kotsadm/pkg/kurl/join_cert.go:182]: i guess `--config` not required here?

File: kotsadm/pkg/kurl/join_cert.go (hunk 4 above)
Comment (divolgin) [reply kotsadm/pkg/kurl/join_cert.go:91]: @marccampbell @emosbaugh I kept this in its own commit intentionally.  We should figure out how to make this work, maybe in 1.19: https://github.com/replicatedhq/kots/commit/6ba2ca6a#diff-d813b91502ac37baa1004104af667747R16

File: kotsadm/pkg/kurl/join_cert.go (hunk 8 above)
Comment (divolgin) [reply kotsadm/pkg/kurl/join_cert.go:182]: nope.  just works.  there must be some defaults that we need to copy to make the Go version of this work.

File: kotsadm/pkg/kurl/configmap.go [hunk 9]
```diff
@@ -69,10 +70,15 @@ func UpdateConfigMap(client kubernetes.Interface, generateBootstrapToken, upload
 	if uploadCerts {
 		certsDuration := time.Hour * 2
 		certsExpiration := time.Now().Add(certsDuration)
-		key, err := createCertAndKey(context.TODO(), client, "default")
+
+		ctx, cancel := context.WithTimeout(context.TODO(), 30*time.Second)
```
Comment (emosbaugh) [inline kotsadm/pkg/kurl/configmap.go:74]: I wonder if you want to give it a little more time?

**Pull request merged**